usr/man/man/usr/share/man/man2/uadmin.2
'\" te
.\" Copyright (c) 2009, Sun Microsystems, Inc.  All Rights Reserved.
.\" Copyright 1989 AT&T
.TH uadmin 2 "25 Mar 2009" "SunOS 5.11" "System Calls"
.SH NAME
uadmin \- administrative control
.SH SYNOPSIS
.LP
.nf
#include <sys/uadmin.h>

\fBint\fR \fBuadmin\fR(\fBint\fR \fIcmd\fR, \fBint\fR \fIfcn\fR, \fBuintptr_t\fR \fImdep\fR);
.fi

.SH DESCRIPTION
.sp
.LP
The \fBuadmin()\fR function provides control for basic administrative functions. This function is tightly coupled to the system administrative procedures and is not intended for general use. The argument \fImdep\fR is provided for machine-dependent use and is not defined here. It should be initialized to \fINULL\fR if not used.
.sp
.LP
As specified by \fIcmd\fR, the following commands are available:
.sp
.ne 2
.mk
.na
\fB\fBA_SHUTDOWN\fR\fR
.ad
.RS 14n
.rt  
The system is shut down. All user processes are killed, the buffer cache is flushed, and the root file system is unmounted. The action to be taken after the system has been shut down is specified by \fIfcn\fR. The functions are generic; the hardware capabilities vary on  specific machines.  
.sp
.ne 2
.mk
.na
\fB\fBAD_HALT\fR\fR
.ad
.RS 17n
.rt  
Halt the processor(s).
.RE

.sp
.ne 2
.mk
.na
\fB\fBAD_POWEROFF\fR\fR
.ad
.RS 17n
.rt  
Halt the processor(s) and turn off the power.
.RE

.sp
.ne 2
.mk
.na
\fB\fBAD_BOOT\fR\fR
.ad
.RS 17n
.rt  
Reboot the system, using the kernel file.
.RE

.sp
.ne 2
.mk
.na
\fB\fBAD_IBOOT\fR\fR
.ad
.RS 17n
.rt  
Interactive reboot; user is prompted for bootable program name.
.RE

.sp
.ne 2
.mk
.na
\fB\fBAD_FASTREBOOT\fR\fR
.ad
.RS 17n
.rt  
Bypass BIOS and boot loader
.RE

.RE

.sp
.ne 2
.mk
.na
\fB\fBA_REBOOT\fR\fR
.ad
.RS 14n
.rt  
The system stops immediately without any further processing. The action to be taken next is specified by \fIfcn\fR as above.
.RE

.sp
.ne 2
.mk
.na
\fB\fBA_DUMP\fR\fR
.ad
.RS 14n
.rt  
The system is forced to panic immediately without any further processing and a crash dump is written to the dump device (see \fBdumpadm\fR(1M)).  The action to be taken next is specified by \fIfcn\fR, as above.
.RE

.sp
.ne 2
.mk
.na
\fB\fBA_REMOUNT\fR\fR
.ad
.RS 14n
.rt  
The root file system is mounted again after having been fixed. This should be used only during the startup process.
.RE

.sp
.ne 2
.mk
.na
\fB\fBA_FREEZE\fR\fR
.ad
.RS 14n
.rt  
Suspend the whole system.  The system state is preserved in the state file. The following subcommands, specified by \fIfcn\fR, are available.
.sp
.ne 2
.mk
.na
\fB\fBAD_SUSPEND_TO_DISK\fR\fR
.ad
.sp .6
.RS 4n
Save the system state to the state file. This subcommand is equivalent to ACPI state S4.
.RE

.sp
.ne 2
.mk
.na
\fB\fBAD_CHECK_SUSPEND_TO_DISK\fR\fR
.ad
.sp .6
.RS 4n
Check if your system supports suspend to disk. Without performing a system suspend/resume, this subcommand checks if this feature is currently available on your system.
.RE

.sp
.ne 2
.mk
.na
\fB\fBAD_SUSPEND_TO_RAM\fR\fR
.ad
.sp .6
.RS 4n
Save the system state to memory This subcommand is equivalent to ACPI state S3.
.RE

.sp
.ne 2
.mk
.na
\fB\fBAD_CHECK_SUSPEND_TO_RAM\fR\fR
.ad
.sp .6
.RS 4n
Check if your system supports suspend to memory. Without performing a system suspend/resume, this subcommand checks if this feature is currently available on your system.
.RE

The following subcommands, specified by \fIfcn\fR, are obsolete and might be removed in a subsequent release:
.sp
.ne 2
.mk
.na
\fB\fBAD_COMPRESS\fR\fR
.ad
.sp .6
.RS 4n
Save the system state to the state file with compression of data. This subcommand has been replaced by \fBAD_SUSPEND_TO_DISK\fR, which should be used instead.
.RE

.sp
.ne 2
.mk
.na
\fB\fBAD_CHECK\fR\fR
.ad
.sp .6
.RS 4n
Check if your system supports suspend and resume.  Without performing a  system suspend/resume, this command checks if this feature is currently available on your system. This subcommand has been replaced by \fBAD_CHECK_SUSPEND_TO_DISK\fR, which should be used instead.
.RE

.sp
.ne 2
.mk
.na
\fB\fBAD_FORCE\fR\fR
.ad
.sp .6
.RS 4n
Force \fBAD_COMPRESS\fR even when threads of user applications are not suspendable. This subcommand should never be used, as it might result in undefined behavior.
.RE

.RE

.SH RETURN VALUES
.sp
.LP
Upon successful completion, the value returned depends on \fIcmd\fR as follows:
.sp
.ne 2
.mk
.na
\fB\fBA_SHUTDOWN\fR\fR
.ad
.RS 14n
.rt  
Never returns.
.RE

.sp
.ne 2
.mk
.na
\fB\fBA_REBOOT\fR\fR
.ad
.RS 14n
.rt  
Never returns.
.RE

.sp
.ne 2
.mk
.na
\fB\fBA_FREEZE\fR\fR
.ad
.RS 14n
.rt  
\fB0\fR upon resume.
.RE

.sp
.ne 2
.mk
.na
\fB\fBA_REMOUNT\fR\fR
.ad
.RS 14n
.rt  
\fB0\fR.
.RE

.sp
.LP
Otherwise, \fB\(mi1\fR is returned and \fBerrno\fR is set to indicate the error.
.SH ERRORS
.sp
.LP
The \fBuadmin()\fR function will fail if:
.sp
.ne 2
.mk
.na
\fB\fBEBUSY\fR\fR
.ad
.RS 11n
.rt  
Suspend is already in progress.
.RE

.sp
.ne 2
.mk
.na
\fB\fBEINVAL\fR\fR
.ad
.RS 11n
.rt  
The \fIcmd\fR argument is invalid.
.RE

.sp
.ne 2
.mk
.na
\fB\fBENOMEM\fR\fR
.ad
.RS 11n
.rt  
Suspend/resume ran out of physical memory.
.RE

.sp
.ne 2
.mk
.na
\fB\fBENOSPC\fR\fR
.ad
.RS 11n
.rt  
Suspend/resume could not allocate enough space on the root file system to store system information.
.RE

.sp
.ne 2
.mk
.na
\fB\fBENOTSUP\fR\fR
.ad
.RS 11n
.rt  
Suspend/resume is not supported on this platform or the command specified by \fIcmd\fR is not allowed.
.RE

.sp
.ne 2
.mk
.na
\fB\fBENXIO\fR\fR
.ad
.RS 11n
.rt  
Unable to successfully suspend system.
.RE

.sp
.ne 2
.mk
.na
\fB\fBEPERM\fR\fR
.ad
.RS 11n
.rt  
The {\fBPRIV_SYS_CONFIG\fR} privilege is not asserted in the effective set of the calling process.
.RE

.SH ATTRIBUTES
.sp
.LP
See \fBattributes\fR(5) for descriptions of the following attributes:
.sp

.sp
.TS
tab() box;
cw(2.75i) |cw(2.75i) 
lw(2.75i) |lw(2.75i) 
.
ATTRIBUTE TYPEATTRIBUTE VALUE
_
Interface StabilitySee below.
.TE

.sp
.LP
The \fBA_FREEZE\fR command and its subcommands are Committed.
.SH SEE ALSO
.sp
.LP
\fBdumpadm\fR(1M), \fBhalt\fR(1M), \fBkernel\fR(1M), \fBreboot\fR(1M), \fBuadmin\fR(1M), \fBattributes\fR(5), \fBprivileges\fR(5)
.SH WARNINGS
.sp
.LP
Shutting down or halting the system by means of \fBuadmin\fR(1M) does not  update the  boot  archive.  Avoid  using  this command after
.RS +4
.TP
.ie t \(bu
.el o
editing of files such as \fB/etc/system\fR
.RE
.RS +4
.TP
.ie t \(bu
.el o
installing new driver binaries or kernel binaries
.RE
.RS +4
.TP
.ie t \(bu
.el o
updating existing driver binaries or kernel binaries.
.RE
.sp
.LP
Use \fBreboot\fR(1M) or \fBhalt\fR(1M) instead.


usr/src/lib/libc/i386/sys/uadmin.c
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License (the "License").
 * You may not use this file except in compliance with the License.
 *
 * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */

/*
 * Copyright (c) 1995, 2011, Oracle and/or its affiliates. All rights reserved.
 */


/*
 * Wrapper function to implement reboot w/ arguments on x86
 * platforms. Extract reboot arguments and place them in
 * in a transient entry in /[stub]boot/grub/menu.lst
 * All other commands are passed through.
 */
#include "lint.h"
#include "mtlib.h"
#include <fcntl.h>
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/uadmin.h>
#include <unistd.h>
#include <strings.h>
#include <pthread.h>
#include <zone.h>
#include <libscf.h>
#include <thread.h>
#include <dlfcn.h>
#include <atomic.h>

/*
 * Pull in the following three interfaces from libscf without introducing
 * a dependency on it, which since libscf depends on libc would be circular:
 *
 * scf_simple_prop_get
 * scf_simple_prop_next_boolean
 * scf_simple_prop_free
 */
typedef scf_simple_prop_t *(*scf_simple_prop_get_t)(scf_handle_t *,
    const char *, const char *, const char *);
static scf_simple_prop_get_t real_scf_simple_prop_get = NULL;
typedef uint8_t *(*scf_simple_prop_next_boolean_t)(scf_simple_prop_t *);
static scf_simple_prop_next_boolean_t real_scf_simple_prop_next_boolean = NULL;
typedef void (*scf_simple_prop_free_t)(scf_simple_prop_t *);
static scf_simple_prop_free_t real_scf_simple_prop_free = NULL;
static mutex_t scf_lock = DEFAULTMUTEX;

static void
load_scf(void)
{
	void *scf_handle = dlopen("libscf.so.1", RTLD_LAZY);
	scf_simple_prop_get_t scf_simple_prop_get = (scf_handle == NULL)? NULL :
	    (scf_simple_prop_get_t)dlsym(scf_handle, "scf_simple_prop_get");
	scf_simple_prop_next_boolean_t scf_simple_prop_next_boolean =
	    (scf_handle == NULL)? NULL :
	    (scf_simple_prop_next_boolean_t)dlsym(scf_handle,
	    "scf_simple_prop_next_boolean");
	scf_simple_prop_free_t scf_simple_prop_free =
	    (scf_handle == NULL)? NULL :
	    (scf_simple_prop_free_t)dlsym(scf_handle, "scf_simple_prop_free");

	lmutex_lock(&scf_lock);
	if (real_scf_simple_prop_get == NULL ||
	    real_scf_simple_prop_next_boolean == NULL ||
	    real_scf_simple_prop_free == NULL) {
		if (scf_simple_prop_get == NULL)
			real_scf_simple_prop_get = (scf_simple_prop_get_t)(-1);
		else {
			real_scf_simple_prop_get = scf_simple_prop_get;
			scf_handle = NULL;	/* don't dlclose it */
		}
		if (scf_simple_prop_next_boolean == NULL)
			real_scf_simple_prop_next_boolean =
			    (scf_simple_prop_next_boolean_t)(-1);
		else {
			real_scf_simple_prop_next_boolean =
			    scf_simple_prop_next_boolean;
			scf_handle = NULL;	/* don't dlclose it */
		}
		if (scf_simple_prop_free == NULL)
			real_scf_simple_prop_free =
			    (scf_simple_prop_free_t)(-1);
		else {
			real_scf_simple_prop_free = scf_simple_prop_free;
			scf_handle = NULL;	/* don't dlclose it */
		}
		membar_producer();
	}
	lmutex_unlock(&scf_lock);

	if (scf_handle)
		(void) dlclose(scf_handle);
}

static void
check_archive_update(void)
{
	scf_simple_prop_t *prop = NULL;
	boolean_t update_flag = B_FALSE;
	char *fmri = "svc:/system/boot-config:default";
	uint8_t *ret_val = NULL;

	if (real_scf_simple_prop_get == NULL ||
	    real_scf_simple_prop_next_boolean == NULL ||
	    real_scf_simple_prop_free == NULL) {
		load_scf();
	}
	if (real_scf_simple_prop_get == (scf_simple_prop_get_t)(-1) ||
	    real_scf_simple_prop_next_boolean ==
	    (scf_simple_prop_next_boolean_t)(-1) ||
	    real_scf_simple_prop_free == (scf_simple_prop_free_t)(-1)) {
		return;
	}

	prop = real_scf_simple_prop_get(NULL, fmri, "config",
	    "uadmin_boot_archive_sync");
	if (prop) {
		if ((ret_val = real_scf_simple_prop_next_boolean(prop)) !=
		    NULL)
			update_flag = (*ret_val == 0) ? B_FALSE :
			    B_TRUE;
		real_scf_simple_prop_free(prop);
	}

	if (update_flag == B_TRUE)
		(void) system("/usr/sbin/bootadm update-archive");
}

static int
legal_arg(char *bargs)
{
	int i;

	for (i = 0; i < BOOTARGS_MAX; i++, bargs++) {
		if (*bargs == 0 && i > 0)
			return (i);
		if (!isprint(*bargs))
			break;
	}
	return (-1);
}

static char quote[] = "\'";

int
uadmin(int cmd, int fcn, uintptr_t mdep)
{
	extern int __uadmin(int cmd, int fcn, uintptr_t mdep);
	char *bargs, cmdbuf[256];
	struct stat sbuf;
	char *altroot;

	bargs = (char *)mdep;

	if (geteuid() == 0 && getzoneid() == GLOBAL_ZONEID &&
	    (cmd == A_SHUTDOWN || cmd == A_REBOOT)) {
		int off = 0;

		switch (fcn) {
		case AD_IBOOT:
		case AD_SBOOT:
		case AD_SIBOOT:
			/*
			 * These functions fabricate appropriate bootargs.
			 * If bootargs are passed in, map these functions
			 * to AD_BOOT.
			 */
			if (bargs == 0) {
				switch (fcn) {
				case AD_IBOOT:
					bargs = "-a";
					break;
				case AD_SBOOT:
					bargs = "-s";
					break;
				case AD_SIBOOT:
					bargs = "-sa";
					break;
				}
			}
			/*FALLTHROUGH*/
		case AD_BOOT:
		case AD_FASTREBOOT:
			if (bargs == 0)
				break;	/* no args */
			if (legal_arg(bargs) < 0)
				break;	/* bad args */

			/* avoid cancellation in system() */
			(void) pthread_setcancelstate(PTHREAD_CANCEL_DISABLE,
			    NULL);

			/* check for /stubboot */
			if (stat("/stubboot/boot/grub/menu.lst", &sbuf) == 0) {
				altroot = "-R /stubboot ";
			} else {
				altroot = "";
			}

			if (fcn == AD_FASTREBOOT) {
				char *newarg, *head;
				char bargs_scratch[BOOTARGS_MAX];

				bzero(bargs_scratch, BOOTARGS_MAX);

				bcopy(bargs, bargs_scratch, strlen(bargs));
				head = bargs_scratch;
				newarg = strtok(bargs_scratch, " ");

				if (newarg == NULL || newarg[0] == '-')
					break;

				/* First argument is rootdir */
				if (strncmp(&newarg[strlen(newarg)-4],
				    "unix", 4) != 0) {
					newarg = strtok(NULL, " ");
					off = newarg - head;
				}

				/*
				 * If we are using alternate root via
				 * mountpoint or a different BE, don't
				 * bother to update the temp menu entry.
				 */
				if (off > 0)
					break;
			}

			/* are we rebooting to a GRUB menu entry? */
			if (isdigit(bargs[0])) {
				int entry = strtol(bargs, NULL, 10);
				(void) snprintf(cmdbuf, sizeof (cmdbuf),
				    "/usr/sbin/bootadm set-menu %sdefault=%d",
				    altroot, entry);
			} else {
				(void) snprintf(cmdbuf, sizeof (cmdbuf),
				    "/usr/sbin/bootadm -m update_temp %s"
				    "-o %s%s%s", altroot, quote,
				    &bargs[off], quote);
			}
			(void) system(cmdbuf);
		}
		check_archive_update();
	}
	return (__uadmin(cmd, fcn, mdep));
}
